{{extend 'layout.html'}}

<div class="row">
<div class="span2">

<h4>Table of Contents</h4>
<li><a href="#Running">Running predictions</a></li>
<li><a href="#Viewing">Viewing predictions</a></li>
<li><a href="#Genomewide">Genome wide analysis</a></li>
<li><a href="#Conservation">Eptiope conservation</a></li>
<li><a href="#Administration">Administration</a></li>
<li><a href="#Installation">Installation</a></li>
<li><a href="#Programmers">For Programmers</a></li>
<li><a href="#References">References</a></li>
</div>

<div class="span9 offset0">

<h3>Background</h3>
This is a web interface for running and visualising epitope predictions (currently MHC binding) and
allow integrated analysis for entire proteomes. The application is built on the web2py
platform.

<a name="Running predictions"></a>
<h3>Running predictions</h3>

<h4>Submission page</h4>
<p>
The binding predictions are made by submitting jobs via the submission form. Several
prediction methods can be chosen and will be run in turn. Once the submit buttone is
pressed the user is directed to another page indicating the task id and status shown
as QUEUED. The id is unique identifier for the job. Refreshing the page will show the
task progress moving from ASSIGNED to RUNNING then COMPLETED once the job is done.
This page can be retrieved if you know the job id with this URL:
<div class="monospace">
http://localhost/epitopemap/default/jobsubmitted?id=x
</div>

<p>
Jobs are queued and run in the background using the web2py scheduler system and are run
consecutively. However it is possible to create multiple workers when web2py is started and
thus multiple queues. You could do this if you are using a computer with multiple
processors and want to make better use of them.
<p>
The time taken to run predictions depends on the number of alleles, number of proteins and
the prediction method. Tepitope is fastest and will take 15-20 minutes to do a genome with
~4000 proteins for each allele. NetMHCIIpan will be noticeably slower and may take several
hours.

<h4>Prediction methods</h4>
There are many MHC binding (T cell), B cell and other predictors available. Those currently
supported by this application are due to their popularity and reliability. We primarily
focus on MHC binding prediction at this time.
<p>
The Tepitope method is our own Python implementation of TEPITOPEPan and requires no external program
to run. netMHCIIpan must be downloaded separately from the website and installed on your system.
The process is quite simple. The same applies for the IEDB tools. Both of these tools are free
for academic use but we do not distribute them with this app.
Remember that this app is designed as an integrated way to use these tools but
is not meant to replace them in all respects. The netMHCIIpan server at the CBS
should be consulted for some more detailed options on using that program.

Here are the available methods with download links for external tools:
<p>
<li>
{{=A('TEPITOPEPan', _href='http://www.biokdd.fudan.edu.cn/Service/TEPITOPEpan/TEPITOPEpan.html')}}
(Implemented internally, so NO download required).
</li>
<li>
{{=A('NetMHCIIpan', _href='http://www.cbs.dtu.dk/services/NetMHCIIpan/')}}
</li>
<li>{{=A('IEDB MHCI tools', _href='http://tools.immuneepitope.org/mhci/download/')}}</li>

<li>{{=A('IEDB BCell tools', _href='http://tools.immuneepitope.org/bcell/download/')}} </li>

<h4>Selection of alleles</h4>
<p>
Selection of which alleles to run will be entirely dependent upon the end results required. Users
must have a particular goal in mind when making predictions. Simply using all alleles will NOT
be helpful and will be very slow since there are so many human alleles.
<p>
Some alleles are not available for all methods and will be ignored for those methods.
MHCII DP and DQ alleles are for netMHCIIpan only.

<h4>How predictions are stored</h4>
<p>
Storage of results uses a very simple method. Results run via the web app are simply added
to the database recording an identifier. All results are then stored according to the label/identifier
with one directory per genome and then a folder for each prediction method. Inside this is a file
for each protein prediction as follows.  Each file is stored  in
{{=A('MessagePack', _href='http://msgpack.org')}} format, which is a binary format for
efficient storage). Data is stored in this way so that results can be quickly loaded for a
specific protein.

Example:

<div class="monospace">
<p>../static/results/identifier/genome/method/locus_tag.mpk
<p>e.g:
<p>../static/results/testresults/Mbovis/tepitope/Mb0012.mpk
</div>
<br>
This example is for tepitope prediction data for the protein Mb0012 in Mbovis. All results
are stored in the static/results folder of the web application.

<h4>Adding sequences</h4>
<p>
Before predictions you need to add your proteins/proteome to the application.
Annotated protein sequences are added as either fasta, embl or genbank format.
Usually these will represent the annotated sequences of a species but any properly
formatted file can be used that groups the proteins together.
Genbank format files for an organism can be found on the
 {{=A('NCBI genomes page',_href='http://www.ncbi.nlm.nih.gov/genome/')}} or downloaded
{{=A('via ftp',_href='ftp://ftp.ncbi.nih.gov/genomes/')}}.

<p>
The program will use the locus tags in the genbank file for each protein. If these are not
available the gene names are used as locus tags. For fasta inputs the identifier following the ">" is used.
<b>Genbank or embl format is recommended</b> since fasta files can have non-standard or confusing
ids like gi|31563518|ref|NP_852610.1| which is not very user friendly. Fasta files do not contain
features present in genbank format such as gene product or location that can be useful for
identifying the protein.
Improperly formatted files may not work.

<a name="Viewing"></a>
<h3>Viewing predictions</h3>
<p>
Every set of predictions for a genome (or maybe just a group of protein sequences) has an identifier
associated with it. This is simply a string identifying a set of binding predictions run with the
same parameters, e.g. a specific set of alleles. This 'prediction id' simply groups your predictions
together in the same folder and can be helpful to separate your predictions from another set if
you are working on the same proteins. So if you want a different set of predictions for the
same proteins then create a new label/id.

<h4>Quick search</h4>
This form on the home page allows you to get a quick preview and see if a particular
protein is present and has some predictions made. If there are results available a small
plot will be displayed in the page. Either the locus tag or gene name can be
used but only the tag is unique. If there are proteins with the same gene name in the source
annotation then only the first one is used here. The gene name overrides the locus tag in this
form.

<h4>Quick view</h4>
This is the usual interface for viewing results. It consists of a form on the left side which
updates the results on the right panel of the page when the 'update' button is pressed.
The id,genome and required protein tag must all be entered correctly.

<h4>Binding prediction plots</h4>
<p>
Predictors will usually make multiple scoring predictions for each allele for an n-mer set of binders.
 This information can be hard to display in a single plot. The purpose of the plot is to allow
the user to quickly visualise where epitopes might be along the sequence and compare all alleles and
 different methods in one plot. The default representation is a set of tracks with bars
 representing predicted binders for each allele ordered by position on the sequence
 as shown below. Plots can be zoomed in and out and panned left right.
<p>
<img src=../static/images/help_plots.png width=900>

<h4>Cut-offs/thresholds</h4>

Selection of predicted binders can be done based on the percentile rank or MHC binding affinity
(or whatever score is a proxy for binding). It has also been shown that absolute binding
affinity threshold correlates better with immunogenicity and MHC allele-specific thresholds
should be used to improve correlations.

For MHC-II the IEDB currently recommends making selections based on a consensus percentile rank of the top 10%.
See {{=A('this iedb help page',_href='http://help.iedb.org/entries/23854373-Selecting-thresholds-cut-offs-for-MHC-class-I-and-II-binding-predictions')}}
for more information.

<h4>Using fixed URIs</h4>
<div class="monospace">
http://localhost/epitopemap/default/protein/results_bovine/Mbovis/Mb0012
</div>
<br>
The protein URIs are structured so that they can be entered in the address bar if required.
However the easiest way to find and view predictions for a known protein name is via the <b>quickview</b>
page which has form for entering the protein details and a few other options.

<h4>Searching for a protein</h4>

<a name="Genomewide"></a>
<h3>Genome wide analysis</h3>
<p>
This page performs an analysis of all proteins to find the top binders found in
n alleles. A global percentage cutoff can also be chosen. This allows a quick overview
of the proteins with highest percentage of binders and those with epitopes that cluster
together. The result is a table that can be sorted according to various metrics
such as the number of binders per unit length of each protein or the number and
density of epitope clusters.
<p>
Explanation of results table columns:
<li>mean: mean score for binders in that protein</li>
<li>size: number of binders found in protein</li>
<li>amax: max score</li>
<li>length: length of protein</li>
<li>order: order in genome annotation (for genbank files)</li>
<li>perc: percentage of binders per protein length</li>
<li>density: highest density epitope cluster found</li>
<p>
<p>
Two additional tables showing the top peptides and clusters are also produced.

<a name="Conservation"></a>
<h3>Epitope conservation analysis</h3>
<p>
This interface allows you to get an estimate of how well conserved a predicted epitope is across a
given set of sequences at a certain identity level.
<p>
Orthologs for the sequence are first retrieved using an online blast. This can take
up to several minutes but results are saved for later use.
Once the alignments are returned the conservation of each epitope can be calculated. This is
defined as <b>the percentage of sequences where the epitope is present out of the total aligned</b>. The
set of sequences can be chosen based on a desired identity cutoff or taxon specification. For
example, the sequences may constitute all orthologs within a bacterial species or just a set of
strains.

<h4>Results</h4>
<p>
Results are given as a table of all binders with >n alleles and their percentage
conservation at that identity level. In addition a plot showing the variation of
this value with cutoff identity gives an indication of how well an epitope is conserved
across species/strains. This will be context dependent.

<h4>BLAST options</h4>
Our method uses the NCBI online Blast service to retrieve orthologous matches. By default
all matches are retrieved with an expect value lower than 10 and the sequences can then be
filtered by percentage identity. It may be preferable or necessary to refine blast queries
to narrow down the sequence set to a specific taxon or species.
<p>
This can be done using the entrez query text box (empty by default). This restricts the
search to a subset of entries from the nr protein database fitting the requirement
of the Entrez query.
For example, to restrict the results to Actinobacteria we would use: <i>txid201174[ORGN]</i>.

See {{=A('this page',_href='http://www.ncbi.nlm.nih.gov/BLAST/blastcgihelp.shtml#entrez_query')}}
for more information.

<p>
The following field names may be useful:
<div class="monospace">
{{x=['ALL, All Fields, All terms from all searchable fields',
 'UID, UID, Unique number assigned to each sequence',
 'FILT, Filter, Limits the records',
 'WORD, Text Word, Free text associated with record',
 'TITL, Title, Words in definition line',
 'KYWD, Keyword, Nonstandardized terms provided by submitter',
 'AUTH, Author, Author(s) of publication',
 'JOUR, Journal, Journal abbreviation of publication',
 'VOL, Volume, Volume number of publication',
 'ORGN, Organism, Scientific and common names of organism, and all higher levels of taxonomy',
 'ACCN, Accession, Accession number of sequence',
 'PACC, Primary Accession, Does not include retired secondary accessions',
 'GENE, Gene Name, Name of gene associated with sequence',
 'PROT, Protein Name, Name of protein associated with sequence',
 'ECNO, EC/RN Number, EC number for enzyme or CAS registry number',
 'PDAT, Publication Date, Date sequence added to GenBank',
 'MDAT, Modification Date, Date of last update',
 'SUBS, Substance Name, CAS chemical name or MEDLINE Substance Name',
 'PROP, Properties, Classification by source qualifiers and molecule type',
 'SQID, SeqID String, String identifier for sequence',
 'GPRJ, Genome Project, Genome Project',
 'SLEN, Sequence Length, Length of sequence',
 'FKEY, Feature key, Feature annotated on sequence',
 'ORGL, Organelle, Organelle']}}
{{for i in x:}}
{{=LI(i)}}
{{pass}}
</div>

<a name="Administration"></a>
<h3>Administration</h3>

<h4>Paths to binaries</h4>
The IEDB tools and/or netMHCIIpan distributions are required to be installed locally
for the web application to use them. To ensure that these programs can be found you can
the path to the install folders here.

<h4>Add/remove genomes</h4>
See the instructions for adding sequences above.

<h4>Add/remove predictions</h4>
A row with the identifier is added to the database whenever a set of predictions is submitted
with a new id. Therefore this menu is normally only used to delete old predictions. Old results
should then be deleted manually from the filesystem.

<h4>Starting the server</h4>
The server is started by typing the following command from inside the web2py folder:
<div class="monospace">
sudo python web2py.py -i localhost -a 123 -p 80 -K epitopemap -X
</div>
-i web address
-p is the port
-K specifies the scheduler

<a name="Installation"></a>
<h3>Installation</h3>
<p>
This application has been built and tested on the
{{=A('Ubuntu Linux ',_href='http://www.ubuntu.com/')}}
operating system (version 14.04). web2py and the required Python dependencies are also
available on both Windows and Mac OSX also though these have not been fully tested yet.
We therefore <i> currently recommend running this web app on a linux OS.</i>

Follow these steps:
<li>Install web2py, which can be downloaded
{{=A('here',_href='http://www.web2py.com/init/default/download')}}</li>
<li>Unzip web2py and place in the folder where you wish to run it</li>
<li>Download the epitopemap files
{{=A('here',_href='https://github.com/dmnfarrell/epitopemap/zipball/master')}} </li>
<li>Unzip epitopemap and place under web2py/applications </li>
<li>Start the server as outlined above</li>
<li>Go to http://localhost/epitopemap in your browser</li>
<p>

<h4>Other software</h4>
The ncbi-blast+ tools and muscle are needed for conservation analysis. On Ubuntu type:
<div class="monospace">
sudo apt-get install ncbi-blast+ muscle
</div>

<h4>Python dependencies</h4>
Setup currently requires you to install the Python libraries yourself. This is
a cleaner approach than providing them with the application and is now simple
on linux. In addition we recommend you use easy_install or pip to install the packages
rather than the OS package manager (e.g. apt-get) but both methods should work.
<p>
On Ubuntu type the following on the command line to install the Python modules:
<div class="monospace">
sudo apt-get install python-pip
<br>
sudo pip install numpy pandas matplotlib biopython bokeh mpld3
</div>

<a name="Programmers"></a>
<h3>For Programmers</h3>
<h4>Python mhcpredict library</h4>
The Python Predictor class and other modules used to implement the prediction routines are available
at {{=A('this github page',_href='https://github.com/dmnfarrell/mhcpredict')}}. It is possible
to implement other prediction methods using these classes and add them to the web application.
See instructions on the github page.

<a name="References"></a>

<h3>References</h3>
The following papers provide some background to the field:

<li>
{{=T('''The utility and limitations of current web-available algorithms to predict
peptides recognized by CD4 T cells in response to pathogen infection.
Chaves et al. J Immunol. May 1, 2012; 188(9): 4235–4248.''')}}
{{=A('link',_href='http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3331894')}}
</li>

<li>
{{=T('''Predictions versus high-throughput experiments in T-cell epitope discovery:
competition or synergy? Expert Review Vaccines, 11(1), 43–54. Lundegaard et al. 2012.''')}}
{{=A('link',_href='http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3297080')}}
</li>

<li>
{{=T('''Development of an epitope conservancy analysis tool to facilitate the design
 of epitope-based diagnostics and vaccines. BMC bioinformatics 8, 361 (2007). Bui et al. ''')}}
{{=A('link',_href='http://www.biomedcentral.com/1471-2105/8/361')}}
</li>

<li>
{{=T('''HLA class I alleles are associated with peptide-binding repertoires of different size,
affinity, and immunogenicity.  Journal of Immunology 2013, 191(12):5831-5839. Paul et al.''')}}
{{=A('link',_href='http://www.ncbi.nlm.nih.gov/pubmed/24190657')}}
</li>

</div>
</div>


